File: DynammicProgramming/charliAndPilots/charliAndPilots.cpp
#include<iostream>
#include<climits>



int minSalary(int* captain, int* assistant, int n, int X) {
    if (n == 0) {
        return 0;
    }

    // if x is 0, that means either we create pairs of assistant and captain for current n or we start from begining
    int ans{};
    if (X == 0) {
        // x is 0, then current pilot is consider as younger, so he can't become captain, but he can become assistant
        ans = minSalary(captain+1, assistant+1, n-1, X+1) + assistant[0];
    } else if (X == n) {
        ans = minSalary(captain+1, assistant+1, n-1,X-1) + captain[0];
    } else {
        int option1 = minSalary(captain+1,assistant+1,n-1,X+1) + assistant[0];
        int option2 = minSalary(captain+1, assistant+1,n-1,X-1) + captain[0];
        ans = std::min(option1,option2);
    }
    return ans;
}


int minSalary(int* captain, int* assistant, int n, int X, int** dp) {
    if (n == 0) {
        return 0;
    }

    if(dp[n][X] != -1) {
        return dp[n][X];
    }

    // if x is 0, that means either we create pairs of assistant and captain for current n or we start from begining
    int ans{};
    if (X == 0) {
        // x is 0, then current pilot is consider as younger, so he can't become captain, but he can become assistant
        ans = minSalary(captain+1, assistant+1, n-1, X+1,dp) + assistant[0];
    } else if (X == n) {
        ans = minSalary(captain+1, assistant+1, n-1,X-1,dp) + captain[0];
    } else {
        int option1 = minSalary(captain+1,assistant+1,n-1,X+1,dp) + assistant[0];
        int option2 = minSalary(captain+1, assistant+1,n-1,X-1,dp) + captain[0];
        ans = std::min(option1,option2);
    }
    dp[n][X] = ans;
    return ans;
}



int main(){
    int n;
    std::cin >> n;
    int captain[n],assistant[n];
    for (int i{}; i<n; i++) {
        int x,y;
        std::cin >> x >> y;
        captain[i] = x;
        assistant[i] = y;
    }

    int ans {minSalary(captain,assistant,n,0)};
    std::cout << ans << std::endl;

    int** dp {new int*[n+1]};
    for(int i{}; i<n+1; i++) {
        dp[i] = new int[n];
        for(int j{}; j<n; j++) {
            dp[i][j] = -1;
        }
    }

    ans = minSalary(captain,assistant,n,0,dp);
    std::cout << ans << std::endl;
}

File: DynammicProgramming/charliAndPilots/charliAndPilots.js
(function(){


    function minSalary(captain, assistant, n, start, X) {
        if (start === n) {
            return 0;
        }

        let ans;
        if (X === 0) {
            ans = minSalary(captain,assistant,n,start+1, X+1) + assistant[start];
        } else if (X === n - start) {
            ans = minSalary(captain,assistant,n,start+1, X-1) + captain[start];
        } else {
            let option1 = minSalary(captain,assistant,n,start+1,X+1) + assistant[start];
            let option2 = minSalary(captain,assistant,n,start+1, X-1) + captain[start];
            ans = Math.min(option1, option2);
        }
        return ans;
    }


    (function main(){

        let n = 4;
        let captain = [5000,6000,8000,9000];
        let assistant = [3000,2000,1000,6000];

        let ans = minSalary(captain,assistant,n,0,0);
        console.log(ans);

    }());



}());